v01-object-or-null/src/main/java/dev/futa/tutorial/pesel/v01/PeselService.java

```diff
@@ -0,0 +1,106 @@
+package dev.futa.tutorial.pesel.v01;
+
+import com.google.common.collect.ImmutableMap;
+import dev.futa.tutorial.pesel.Gender;
+
+import java.time.DateTimeException;
+import java.time.LocalDate;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import static dev.futa.tutorial.pesel.Gender.FEMALE;
+import static dev.futa.tutorial.pesel.Gender.MALE;
+import static java.lang.Character.getNumericValue;
+import static java.lang.Integer.parseInt;
+
+public class PeselService {
+
+  private static final byte PESEL_LENGTH = 11;
+  private static final int[] WEIGHTS_PESEL = {1, 3, 7, 9, 1, 3, 7, 9, 1, 3, 1};
+  private static final Pattern PESEL_PATTERN = Pattern.compile("[0-9]{11}");
+
+  private static final Map<Integer, Integer> YEAR_CODES =
+      ImmutableMap.of(
+          0, 1900,
+          20, 2000,
+          40, 2100,
+          60, 2200,
+          80, 1800);
+
+  /**
+   * Method decodes date od birth and gender from PESEL number
+   *
+   * @param pesel PESEL number to decode data from
+   * @return object contains decoded data and PESEL itself. Will return null if PESEL do not meet
+   *     validation rules or coded date is not valid date.
+   */
+  public PeselInfo decodePesel(final String pesel) {
+
+    if (pesel == null) {
+      return null;
+    }
+
+    if (isPatternValid(pesel) && isChecksumValid(pesel)) {
+      final LocalDate birthDate = extractBirthDate(pesel);
+      final Gender gender = extractGender(pesel);
+
+      if (birthDate != null) {
+        return new PeselInfo(pesel, birthDate, gender);
+      }
+    }
+
+    return null;
+  }
+
+  private boolean isChecksumValid(String pesel) {
+    final int[] digitsOfPesel = peselToDigitsArray(pesel);
+    return calculateWeightedSum(digitsOfPesel) % 10 == 0;
```

Comment:
tu by mozna to zrobic bardziej funkcyjnie, korzystajac z `IntStream`

Reply:
To jest implementacje prymitywna z pełną świadomością tego faktu. W następnych wersjach, wykorzystujących _Optional_ czy _Either_ będę zmieniał implementację. Tutaj chcę postawić na _prostotę_.

Reply:
Tablica z pętlą wystarczy :-)